librad/src/git/storage/read.rs

```diff
@@ -0,0 +1,347 @@
+            .map(Some)
+            .or_matches(is_not_found_err, || Ok(None))
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn references<'a>(
+        &'a self,
+        reference: &Reference<Many>,
+    ) -> Result<impl Iterator<Item = Result<git2::Reference<'a>, Error>> + 'a, Error> {
+        self.references_glob(glob::RefspecMatcher::from(RefspecPattern::from(reference)))
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn reference_names<'a>(
+        &'a self,
+        reference: &Reference<Many>,
+    ) -> Result<impl Iterator<Item = Result<ext::RefLike, Error>> + 'a, Error> {
+        self.reference_names_glob(glob::RefspecMatcher::from(RefspecPattern::from(reference)))
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn references_glob<'a, G: 'a>(
+        &'a self,
+        glob: G,
+    ) -> Result<impl Iterator<Item = Result<git2::Reference<'a>, Error>> + 'a, Error>
+    where
+        G: Pattern + Debug,
+    {
+        Ok(self
+            .backend
+            .references()?
+            .filter_map(move |reference| match reference {
+                Ok(reference) => match reference.name() {
+                    Some(name) if glob.matches(name) => Some(Ok(reference)),
+                    _ => None,
+                },
+
+                Err(e) => Some(Err(e.into())),
+            }))
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn reference_names_glob<'a, G: 'a>(
+        &'a self,
+        glob: G,
+    ) -> Result<impl Iterator<Item = Result<ext::RefLike, Error>> + 'a, Error>
+    where
+        G: Pattern + Debug,
+    {
+        let iter = ReferenceNames {
+            iter: self.backend.references()?,
+        };
+        Ok(iter.filter_map(move |refname| match refname {
+            Ok(reflike) if glob.matches(Path::new(reflike.as_str())) => Some(Ok(reflike)),
+            Ok(_) => None,
+
+            Err(e) => Some(Err(e)),
+        }))
+    }
+
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub fn blob<'a>(
+        &'a self,
+        reference: &'a Reference<One>,
+        path: &'a Path,
+    ) -> Result<Option<git2::Blob<'a>>, Error> {
+        ext::Blob::Tip {
+            branch: reference.into(),
+            path,
+        }
+        .get(self.as_raw())
+        .map(Some)
+        .or_matches(|e| matches!(e, blob::Error::NotFound(_)), || Ok(None))
+    }
+
+    pub fn config(&self) -> Result<Config<PhantomData<!>>, Error> {
+        Ok(Config::try_from(&self.backend)?)
+    }
+
+    pub(in crate::git) fn identities<'a, T: 'a>(&'a self) -> Identities<'a, T> {
+        Identities::from(self.as_raw())
+    }
+
+    // TODO: we would need to wrap a few more low-level git operations (such as:
+    // create commit, manipulate refs, manipulate config) in order to be able to
+    // model "capabilities" in terms of traits.
+    pub(in crate::git) fn as_raw(&self) -> &git2::Repository {
```

Comment:
Perhaps this can be `pub(super)`, so it’s not possible to get an almighty handle without having a signer?